File: native/managers/map/level_layer.py
"""
关卡图层
使用组合模式，集成 TileExtractor 和 LevelLayerRenderer
"""

import pygame
from typing import Dict, List, Tuple, Optional, Any

from .map_layer import MapLayer
from .tile_extractor import TileExtractorMixin
from .level_layer_renderer import LevelLayerRenderer


class LevelLayer(MapLayer, TileExtractorMixin):
    """
    关卡图层（组合模式）
    渲染墙壁、障碍物、目标区域、监狱等
    """

    def __init__(self, tiles_image: pygame.Surface, map_width: int,
                 map_height: int, tile_size: int = 32):
        """
        初始化关卡图层

        Args:
            tiles_image: tiles.png 图片 Surface
            map_width: 地图宽度（格子数）
            map_height: 地图高度（格子数）
            tile_size: 瓦片大小
        """
        # 初始化 TileExtractor
        self.init_tile_extractor(tiles_image, tile_size, tile_size)

        self.map_width = map_width
        self.map_height = map_height
        self.walls: List[Dict[str, Any]] = []

        # 创建渲染器
        self._renderer = LevelLayerRenderer(
            self._extract_tile, tile_size, tile_size
        )

    def set_walls(self, walls: List[Dict[str, Any]]):
        """设置墙壁列表"""
        self.walls = walls

    def render(self, surface: pygame.Surface, offset_x: int = 0, offset_y: int = 0):
        """渲染图层（由外部调用具体渲染方法）"""
        pass

    def render_walls(self, surface: pygame.Surface, offset_x: int = 0, offset_y: int = 0):
        """渲染墙壁"""
        self._renderer.render_walls(surface, self.walls, offset_x, offset_y)

    def render_targets(self, surface: pygame.Surface,
                       left_target: List[Tuple[int, int]],
                       right_target: List[Tuple[int, int]],
                       offset_x: int = 0, offset_y: int = 0):
        """渲染目标区域"""
        self._renderer.render_targets(surface, left_target, right_target, offset_x, offset_y)

    def render_prisons(self, surface: pygame.Surface,
                       left_prison: List[Tuple[int, int]],
                       right_prison: List[Tuple[int, int]],
                       offset_x: int = 0, offset_y: int = 0):
        """渲染监狱区域"""
        self._renderer.render_prisons(surface, left_prison, right_prison, offset_x, offset_y)

    def render_obstacles(self, surface: pygame.Surface,
                         obstacles1: List[Tuple[int, int]],
                         obstacles2: List[Tuple[int, int]],
                         offset_x: int = 0, offset_y: int = 0):
        """渲染障碍物"""
        self._renderer.render_obstacles(surface, obstacles1, obstacles2, offset_x, offset_y)

    def is_wall(self, x: int, y: int) -> bool:
        """检查位置是否是墙"""
        return any(w.get("x") == x and w.get("y") == y for w in self.walls)

    def get_tile_at(self, x: int, y: int) -> Optional[Dict[str, Any]]:
        """获取指定位置的图块信息"""
        if not (0 <= x < self.map_width and 0 <= y < self.map_height):
            return None
        for wall in self.walls:
            if wall.get("x") == x and wall.get("y") == y:
                return {"gid": wall.get("tileId", 0), "x": x, "y": y, "is_wall": True}
        return None

    def update(self, delta_time: int):
        """关卡图层不需要更新"""
        pass

    def destroy(self):
        """销毁图层"""
        self.clear_tile_cache()

File: frontend/src/game/managers/map/LevelLayer.ts
/**
 * LevelLayer - 关卡图层
 * 负责渲染墙壁、监狱、目标区域和障碍物
 */
import Phaser from 'phaser'
import type { Position } from '@/types'
import type { MapLayer } from '../MapManager'

/**
 * 关卡图层
 */
export class LevelLayer implements MapLayer {
  private layer: Phaser.Tilemaps.TilemapLayer
  private walls: Array<Position & { tileId?: number }> = []
  private targetTiles: number[] = [13, 14, 15, 25, 26, 27, 37, 38, 39]
  private prisonTiles: number[] = [97, 98, 99, 109, 110, 111, 121, 122, 123]
  private wallTiles: number[] = [45, 46, 47, 57, 59, 69, 70, 71]
  private tree1Tiles: number[] = [6, 18, 30, 29, 28]
  private tree2Tiles: number[][] = [[4, 16], [5, 17]]

  constructor(
    _scene: Phaser.Scene,
    private map: Phaser.Tilemaps.Tilemap,
    tileset: Phaser.Tilemaps.Tileset,
    mapX: number,
    mapY: number,
    private mapWidth: number,
    mapHeight: number
  ) {
    console.log('[LevelLayer] 创建关卡图层')
    this.layer = this.map.createBlankLayer('level', tileset, mapX, mapY)!
    if (!this.layer) {
      throw new Error('Failed to create level layer')
    }
    this.layer.setDepth(10)
    this.layer.setVisible(true)
    this.layer.fill(0, 0, 0, mapWidth, mapHeight)
  }

  setWalls(walls: Array<Position & { tileId?: number }>): void {
    this.walls = walls
  }

  renderWalls(): void {
    this.walls.forEach(wall => {
      const tile = this.layer.getTileAt(wall.x, wall.y)
      if (tile) {
        tile.index = wall.tileId || this.wallTiles[0]
        const collisionId = (wall.x - 1) * this.mapWidth + wall.y
        this.map.setCollision(collisionId)
      }
    })
  }

  renderPrisons(lTeamPrison: Position[], rTeamPrison: Position[]): void {
    lTeamPrison.forEach((prison, i) => {
      const tile = this.layer.getTileAt(prison.x, prison.y)
      if (tile && i < this.prisonTiles.length) {
        tile.index = this.prisonTiles[i]
      }
    })
    rTeamPrison.forEach((prison, i) => {
      const tile = this.layer.getTileAt(prison.x, prison.y)
      if (tile && i < this.prisonTiles.length) {
        tile.index = this.prisonTiles[i]
      }
    })
  }

  renderTargets(lTeamTarget: Position[], rTeamTarget: Position[]): void {
    lTeamTarget.forEach((target, i) => {
      const tile = this.layer.getTileAt(target.x, target.y)
      if (tile && i < this.targetTiles.length) {
        tile.index = this.targetTiles[i]
      }
    })
    rTeamTarget.forEach((target, i) => {
      const tile = this.layer.getTileAt(target.x, target.y)
      if (tile && i < this.targetTiles.length) {
        tile.index = this.targetTiles[i]
      }
    })
  }

  renderObstacles(obstacles1: Position[], obstacles2: Position[]): void {
    obstacles1.forEach(obs => {
      const tile = this.layer.getTileAt(obs.x, obs.y)
      if (tile) {
        tile.index = Phaser.Math.RND.pick(this.tree1Tiles)
        const collisionId = (obs.x - 1) * this.mapWidth + obs.y
        this.map.setCollision(collisionId)
      }
    })

    obstacles2.forEach(obs => {
      const treeTile = Phaser.Math.RND.pick(this.tree2Tiles)
      const tile1 = this.layer.getTileAt(obs.x, obs.y)
      const tile2 = this.layer.getTileAt(obs.x, obs.y + 1)
      if (tile1 && tile2) {
        tile1.index = treeTile[0]
        tile2.index = treeTile[1]
        tile1.setCollision(true)
        tile2.setCollision(true)
      }
    })
  }

  isWall(x: number, y: number): boolean {
    const tile = this.layer.getTileAtWorldXY(x, y, true)
    if (!tile) return false
    return (
      this.wallTiles.indexOf(tile.index) >= 0 ||
      this.tree1Tiles.indexOf(tile.index) >= 0 ||
      this.tree2Tiles[0].indexOf(tile.index) >= 0 ||
      this.tree2Tiles[1].indexOf(tile.index) >= 0
    )
  }

  getTileAtWorldXY(x: number, y: number): Phaser.Tilemaps.Tile | null {
    return this.layer.getTileAtWorldXY(x, y, true) || null
  }

  render(): void {
    // 图层渲染由具体方法处理
  }

  update(): void {
    // 关卡图层不需要更新
  }

  destroy(): void {
    this.layer.destroy()
  }
}
